File: talks/matlab_vs_python/partial_fourier/partial_fourier.m
%% Imports

%% Load Data
h = matfile('data.mat');
img = h.img;

%% 6/8 Partial Fourier sampling
n = randn(96) + 1j*randn(96);
y = fftshift(fft2(img),1) + 0*n;
%y(73:end,:) = 0;

%% Estimate phase
pad = padarray(hann(48),24);
phs = exp(1j*angle(ifft2(ifftshift(y.*pad,1))));

%% POCS reconstruction
est = zeros(96);
iters = 10;
for i = 1:iters
    est = fftshift(fft2(est.*phs),1);
    est(1:72,:) = y(1:72,:);
    est = max(real(ifft2(ifftshift(est,1)).*conj(phs)),0);   
end

%% Plot reconstruction
figure();
subplot(1,2,1);
imshow(abs(ifft2(ifftshift(y,1))),[0 1],'colormap',jet);
title('Zero-Filled');
subplot(1,2,2);
imshow(est,[0 1],'colormap',jet);
title('POCS recon');

File: talks/matlab_vs_python/partial_fourier/partial_fourier.py
#%% Imports
import h5py
import matplotlib.pyplot as plt
import numpy as np

#%% Load Data
h = h5py.File('data.mat','r')
img = np.transpose(h.get('img')['real']+1j*h.get('img')['imag'])

#%% 6/8 Partial Fourier sampling
n = np.random.randn(96,96) + 1j*np.random.randn(96,96)
y = np.fft.fftshift(np.fft.fft2(img),axes=0) + n
y[72:,:] = 0

#%% Estimate phase
pad = np.pad(np.hanning(48),24,'constant')[:,None]
phs = np.exp(1j*np.angle(np.fft.ifft2(np.fft.ifftshift(y*pad,axes=0))))

#%% POCS reconstruction
est = np.zeros((96,96))
iters = 10
for i in range(iters):
    est = np.fft.fftshift(np.fft.fft2(est*phs),axes=0)
    est[:72,:] = y[:72,:]
    est = np.maximum(np.real(np.fft.ifft2(np.fft.ifftshift(est,axes=0))*np.conj(phs)),0)

#%% Plot reconstruction
_, ax = plt.subplots(1,2)
ax[0].imshow(np.abs(np.fft.ifft2(np.fft.ifftshift(y,axes=0))),vmin=0,vmax=1)
ax[0].set_title('Zero-filled')
ax[1].imshow(est, vmin=0, vmax=1)
ax[1].set_title('POCS recon')